File: 프로그래머스/2/169199. 리코쳇 로봇/리코쳇 로봇.py
from collections import deque

def solution(board):
    answer = 0
    
    row = len(board)
    col = len(board[0])
    
    maps = []
    
    for r in board:
        maps.append(r)
    
    visited= [[-1]*(col) for _ in range (row)]
    
    dy = [-1,1,0,0]
    dx=  [0,0,-1,1]
    
    def bfs (start) :
        y,x = start
        q = deque ([start])
        visited[y][x] =0
        while q :
            cur_y, cur_x = q.popleft()
            
            if maps[cur_y][cur_x] == 'G' :
                return visited[cur_y][cur_x]
            
            for i in range (4) :
                stop_y, stop_x = cur_y, cur_x # stop 멈추는 지점
                
                # next는 다음에 방문할 지점 계산
                # 방문할 지점이 장애물 or  범위 밖
                # break
                # 방문 못하므로 탈출하고, 이전 지점 stop 에서 visited 값 계산
                while True : 
                    next_y = stop_y +dy[i]
                    next_x = stop_x +dx[i]
                    
                    if 0>next_y or next_y >=row  or next_x <0 or next_x >=col :
                        break
                        
                    if maps[next_y][next_x] == 'D' :
                        break
                        
                    stop_y = next_y
                    stop_x = next_x
                
                if visited[stop_y][stop_x] == -1 :
                    visited[stop_y][stop_x] = visited[cur_y][cur_x] +1 # 미끄러지는건 계산 안하므로, 처음 꺼낸 좌표값에서 +1
                    q.append ((stop_y, stop_x))

    # 시작지점 
    for i in range (0, row) :
        for j in range (0,col):
            if visited[i][j] ==-1 and maps[i][j] == 'R':
                answer = bfs((i,j))
    
    if not answer :
        return -1
    else :
        return answer
  

File: 프로그래머스/2/169199. 리코쳇 로봇/리코쳇 로봇.java
import java.util.*;

class Solution {
     
    int [] dy = {-1,1,0,0};
    int [] dx= {0,0,-1,1};
    int answer =0;
    
    public int solution(String[] board) {

        int row= board.length;
        int col= board[0].length();
        int [][] distance = new int [row][col];
        char [][] maps=  new char [row][col];
        
        for (int i=0; i< row; i++) {
            maps[i] = board[i].toCharArray(); // 이게 삽입됨? 되네
        }

        // cur_y, cur_x // next_y, next_x 다음꺼가 D거나 범위를 벗어나는 경우 방향을 바꿔서 이동
        
        for(int [] temp : distance) {
            Arrays.fill (temp, -1);
        }
        
        ArrayDeque <int []> q = new ArrayDeque<>();
        int start_y = 0;
        int start_x = 0;
        
        
        for (int i=0; i< row; i++) {
            for (int j =0; j<col; j++) {
                if (maps[i][j] == 'R') {
                    start_y = i;
                    start_x = j;
                    break;
                }
            }
        }
        distance[start_y][start_x] = 0;
        q.offer(new int [] {start_y, start_x});
        
        while (!q.isEmpty()) {

            int[] cur_cord= q.poll();
            int cur_y =cur_cord[0];
            int cur_x = cur_cord[1];
            
                
            if (maps[cur_y][cur_x] == 'G') {
                return distance[cur_y][cur_x];
            }
            
            for (int i =0; i<4; i++) {
                int next_y = cur_y;
                int next_x = cur_x;
            
            
                while (true) {
                    int test_y = next_y +dy[i];
                    int test_x = next_x + dx[i];

                    if (0>test_y || test_y >= row || test_x <0 || test_x >=col || maps[test_y][test_x] =='D') { //범위 밖이라면
                        break;                  
                    }

                    next_y = test_y;
                    next_x = test_x;

                }

                // 방문처리

                if (distance[next_y][next_x] == -1) {
                    distance[next_y][next_x] = distance[cur_y][cur_x] + 1;
                    q.offer(new int[] {next_y, next_x});
                }

        }
    }
       
        return -1;
    }

}